### PR TITLE
Dragonball: introduce upcall

### PR DESCRIPTION
**Pull request by studychao**
Upcall is a direct communication tool between VMM and guest developed upon vsock. The server side of the upcall is a driver in the guest kernel (kernel patches are needed for this feature) and will start to serve the requests after the kernel starts. And the client side is in Dragonball VMM , it'll be a thread that communicates with vsock through uds.

We want to keep the lightweight of the VM through the implementation of the upcall, through which we could achieve vCPU hotplug, virtio-mmio hotplug without implementing complex and heavy virtualization features such as ACPI virtualization.

fixes: #5642

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

---

This PR is still a WIP since we need to integrate those kernel patches with Kata guest os generation tool and Kata CI. Those work will finish in following days.

- [x] Only apply those as an experimental dragonball patch
- [x] Create a new runtime class for using this patched kernel
may not need a new runtime class and reason explained below.
- [x] Add the "experimental" kernel to what we distribute with Kata Deploy.
issue #5859 tracks this.

### CODE REVIEW

**Comment by fidencio** [comment]
@studychao, please, when working on the integration of those, I'd suggest to:
1. Only apply those as an experimental dragonball patch
2. Create a new runtime class for using this patched kernel
3. Add the "experimental" kernel to what we distribute with Kata Deploy.

**Comment by studychao** [comment]
> @studychao, please, when working on the integration of those, I'd suggest to:
> 
> 1. Only apply those as an experimental dragonball patch
> 2. Create a new runtime class for using this patched kernel
> 3. Add the "experimental" kernel to what we distribute with Kata Deploy.

Thanks @fidencio ! I'll take those advice in mind while doing integration work:)

**Comment by studychao** [comment]
Hi @fidencio , I have finished the most part of the integration work and please help to review the PR.
For your three suggestions, 

> 1.Only apply those as an experimental dragonball patch

I have added upcall patches support into build_kernel.sh and only when `-e` and `-t dragonball` are both enabled which is included in this PR

> 2. Create a new runtime class for using this patched kernel

This one I am not quite get it, where should I add a new runtime class for upcall guest kernel?

> 3.Add the "experimental" kernel to what we distribute with Kata Deploy.

I'll do this in a separate PR for Kata Deploy

Thanks for your review:)

**Comment by studychao** [comment]
PTAL :) @ZizhengBian @liubin @bergwolf @jiangliu @egernst

**Comment by studychao** [comment]
As discussed with @fidencio , he provides an example for runtimeclass (https://github.com/kata-containers/kata-containers/pull/5785/files). Thanks!

However, I think upcall may not be a suitable choice for another runtimeclass.
The reason is that in the previous PR, no matter QEMU with or without Nydus, the Kata Container itself is complete in function (as far as I know, Nydus is a performance bouns) . But upcall provides device hotplug for Dragonball which is a fundamental function so we think it's not a good choice to provide two version of Kata Containers + Dragonball with or without upcall which will cause the Dragonball without upcall one loses the device hotplug function.
So we would like to provide only one runtimeclass with upcall ;)

**Comment by studychao** [comment]
As for the " Add the "experimental" kernel to what we distribute with Kata Deploy." requirement, we need to first add upcall into the static tarball Kata delivered in order to use Kata Deploy. But right now I think this PR focus on introducing upcall to the Kata , and later PRs () could use upcall to enable cpu hotplug, block device hotplug and etc. So we would like to add upcall into Kata and test it stable in CI for doing device hotplug jobs, and then we will add upcall guest kernel to the tarball.

I will create an issue for add upcall guest kernel into static tarball to track this .

File: tools/packaging/kernel/build-kernel.sh [hunk 1]
```diff
@@ -538,6 +538,16 @@ main() {
 
 	[ -z "${subcmd}" ] && usage 1
 
+	if [[ ${build_type} == "experimental" ]] && [[ ${hypervisor_target} == "dragonball" ]]; then
+		build_type="dragonball-experimental"
+		if [ -n "$kernel_version" ];  then
+			kernel_ver=$(get_major_kernel_version "${version}")
```
Comment (fidencio) [inline tools/packaging/kernel/build-kernel.sh:544]:
```suggestion
			kernel_ver=$(get_major_kernel_version "${kernel_version}")
```

Comment (fidencio) [reply tools/packaging/kernel/build-kernel.sh:544]: Then here you declare `kernel_ver` but does not actually use it, which makes me thing that ....

File: tools/packaging/kernel/build-kernel.sh [hunk 2]
```diff
@@ -538,6 +538,16 @@ main() {
...
+			if [[ ${kernel_version} != "5.10" ]]; then
```
Comment (fidencio) [inline tools/packaging/kernel/build-kernel.sh:545]: Here you actually would like to check
```suggestion
			if [ "${kernel_ver}" != "5.10" ]; then
```

**Review by fidencio — COMMENTED** [review]
@studychao, I've left some comments, please, take a look at them.

File: tools/packaging/kernel/build-kernel.sh (hunk 2 above)
Comment (studychao) [reply tools/packaging/kernel/build-kernel.sh]: good catch!!! I've changed `kernel_ver` to `kernel_major_version` and also `if [[ ${kernel_version} != "5.10" ]]; then` to `if [[ ${kernel_major_version} != "5.10" ]]; then`

File: tools/packaging/kernel/build-kernel.sh (hunk 1 above)
Comment (studychao) [reply tools/packaging/kernel/build-kernel.sh]: good catch!!! I've changed kernel_ver to kernel_major_version and also if [[ ${kernel_version} != "5.10" ]]; then to if [[ ${kernel_major_version} != "5.10" ]]; then

**Comment by studychao** [comment]
Hi @fidencio , I have modified the PR accordin to your advice. 
Please take a look at them :)

**Review by fidencio — APPROVED** [review]
lgtm, thanks @studychao!

File: tools/packaging/kernel/patches/5.10.x/dragonball-experimental/0002-upcall-introduce-device-manager-upcall-service.patch [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+After a request is sent to device manager service, db_devmgr_handler
+will start handle the request. A kthread `db_devmgr_server` will be
+created and it will send CONNECT message to the client side to notify
+the client start sending message for device management operations.
+`db_devmgr_process` will be used for determining which device operations
+will be triggered through msg_type. `get_action` will find out the
+action for dealing with the operation and `action` fn will execute the
+actual device management operation in the device manager service.
+
+Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>
```
Comment (ZizhengBian) [inline tools/packaging/kernel/patches/5.10.x/dragonball-experimental/0002-upcall-introduce-device-manager-upcall-service.patch:20]: The sob here is incomplete, right?

Comment (studychao) [reply tools/packaging/kernel/patches/5.10.x/dragonball-experimental/0002-upcall-introduce-device-manager-upcall-service.patch:22]: Thanks, it's added.

**Comment by ZizhengBian** [comment]
LGTM

**Review by ZizhengBian — APPROVED** [review]

**Review by liubin — APPROVED** [review]

**Review by bergwolf — APPROVED** [review]

**Pull request merged**